### PR TITLE
Remove java.lang.reflect.ReflectPermission "suppressAccessChecks"

### PR DESCRIPTION
**Pull request by rmuir**
This is an extremely dangerous permission, basically as dangerous as Unsafe. In addition to this, by removing this permission we solve the main hurdle for the ability to protect confidential data, e.g. `private` fields are really private and so on. It is the most important step we can make to improve the security of the system right now.

Unfortunately its a doozie to get rid of. This pr includes/builds on:
- cleanups of randomized runner and junit4 (a snapshot for now, while dawid is on vacation)
- cleanups to lucene-test-framework
- improvements to securemock to support this: spy/interception support in particular

All of ES is clean but there were some problems with cloud plugins (AWS and GCE). We should fix the code of these third party libraries so we can help protect credentials to their systems!

In order to make them work, I added an "insecure plugin" mechanism where we (list is managed in es-core) can apply additional permissions as workaround to this kind of thing. It is not an open invitation to bring shitty code into our codebase that needs tons of permissions, but a mechanism we can use for problems like this.

As a followup (this PR is crazy enough), I would like to:
- fix AWS and GCE and remove their hacks (we can keep the mechanism)
- cleanup a bunch of other easier permissions we probably no longer need (due to lucene and testrunner cleanups)
- think about warning or confirmation in pluginmanager when you install such a plugin.

### CODE REVIEW

File: core/src/test/java/org/elasticsearch/bootstrap/BootstrapForTesting.java [hunk 1]
```diff
@@ -83,6 +83,13 @@
         if (systemPropertyAsBoolean("tests.security.manager", true)) {
             try {
                 Security.setCodebaseProperties();
+                // if its an insecure plugin, its not easy to simulate here, since we don't have a real plugin install.
+                // we just do our best so unit testing can work. integration tests for such plugins are essential.
+                String artifact = System.getProperty("tests.artifact");
+                String insecurePluginProp = Security.INSECURE_PLUGINS.get(artifact);
+                if (insecurePluginProp != null) {
+                    System.setProperty(insecurePluginProp, "file:/-");
```
Comment (s1monw) [inline core/src/test/java/org/elasticsearch/bootstrap/BootstrapForTesting.java]: out of curiosity what does ` "file:/-` mean?

**Comment by s1monw** [comment]
LGTM can you put a note / reference to the RandomizedRunner issue that explains why we reference a snapshot I think it's non obvious for others

File: core/src/test/java/org/elasticsearch/bootstrap/BootstrapForTesting.java (hunk 1 above)
Comment (rmuir) [inline core/src/test/java/org/elasticsearch/bootstrap/BootstrapForTesting.java]: It means everything! So the comment is really true, proper integration tests are essential. 

Our unit testing framework for plugins (loading via classpath) is just not good enough. And when that runs the plugin's code and third party dependencies are spread everywhere so we cant formulate a URL pattern for it.

I don't think we should spend time making security stuff complicated or elaborate for tests that will never be realistic anyway. Better to spend the effort fixing the problematic third party libs?

**Comment by rmuir** [comment]
I just backported this to 2.x

**Pull request closed without merging**